src/app/rpmostree-compose-builtin-tree.cxx

```diff
@@ -1430,3 +1441,164 @@ rpmostree_compose_builtin_tree (int             argc,
+    {
+      rpmostree_usage_error (context, "--output-dir must be specified", error);
+      return FALSE;
+    }
+
+  const char *treefile_path = argv[1];
+  const char *extensions_path = argv[2];
+
+  g_autofree char *basearch = rpm_ostree_get_basearch ();
+  g_autoptr(RORTreefile) treefile = ror_treefile_new (treefile_path, basearch, -1, error);
+  if (!treefile)
+    return glnx_prefix_error (error, "Failed to load treefile");
+
+  g_autoptr(OstreeRepo) repo = ostree_repo_open_at (AT_FDCWD, opt_repo, cancellable, error);
+  if (!repo)
+    return FALSE;
+
+  /* this is a similar construction to what's in rpm_ostree_compose_context_new() */
+  g_auto(GLnxTmpDir) cachedir_tmp = { 0, };
+  glnx_autofd int cachedir_dfd = -1;
+  if (opt_cachedir)
+    {
+      if (!glnx_opendirat (AT_FDCWD, opt_cachedir, TRUE, &cachedir_dfd, error))
+        return glnx_prefix_error (error, "Opening cachedir");
+    }
+  else
+    {
+      if (!glnx_mkdtempat (ostree_repo_get_dfd (repo),
+                           "tmp/rpm-ostree-compose.XXXXXX", 0700,
+                           &cachedir_tmp, error))
+        return FALSE;
+
+      cachedir_dfd = fcntl (cachedir_tmp.fd, F_DUPFD_CLOEXEC, 3);
+      if (cachedir_dfd < 0)
+        return glnx_throw_errno_prefix (error, "fcntl");
+    }
+
+  g_autofree char *base_rev = NULL;
+  if (!ostree_repo_resolve_rev (repo, opt_extensions_base_rev, FALSE, &base_rev, error))
+    return FALSE;
+
+  g_autoptr(GVariant) commit = NULL;
+  if (!ostree_repo_load_commit (repo, base_rev, &commit, NULL, error))
+    return FALSE;
+
+  g_autoptr(GPtrArray) packages =
+      rpm_ostree_db_query_all (repo, opt_extensions_base_rev, cancellable, error);
+  if (!packages)
+      return FALSE;
+
+  auto packages_mapping = std::make_unique<rust::Vec<rpmostreecxx::StringMapping>>();
+  for (guint i = 0; i < packages->len; i++)
+    {
+      RpmOstreePackage *pkg = (RpmOstreePackage*)packages->pdata[i];
+      const char *name = rpm_ostree_package_get_name (pkg);
+      const char *evr = rpm_ostree_package_get_evr (pkg);
+      packages_mapping->push_back(rpmostreecxx::StringMapping {k: name, v: evr});
+    }
+
+  auto extensions = rpmostreecxx::extensions_load (extensions_path, basearch, *packages_mapping);
+
+  g_autoptr(RpmOstreeContext) ctx =
+      rpmostree_context_new_tree (cachedir_dfd, repo, cancellable, error);
+  if (!ctx)
+      return FALSE;
+
+  { int tf_dfd = ror_treefile_get_dfd (treefile);
+    g_autofree char *abs_tf_path = glnx_fdrel_abspath (tf_dfd, ".");
+    dnf_context_set_repo_dir (rpmostree_context_get_dnf (ctx), abs_tf_path);
+  }
+
+#define TMP_EXTENSIONS_ROOTFS "rpmostree-extensions.tmp"
+
+  if (!glnx_shutil_rm_rf_at (cachedir_dfd, TMP_EXTENSIONS_ROOTFS, cancellable, error))
+    return FALSE;
+
+  g_print ("Checking out %.7s... ", base_rev);
+  OstreeRepoCheckoutAtOptions opts = { .mode = OSTREE_REPO_CHECKOUT_MODE_USER };
+  if (!ostree_repo_checkout_at (repo, &opts, cachedir_dfd, TMP_EXTENSIONS_ROOTFS,
```

Comment:
Sort of tangentially related discussion...do you see us running this in cosa via supermin or not?  I think that influences checkouts of the tree because if we don't do supermin then the cache ostreerepo may be on the `cache.qcow2`.  Perhaps the simplest is to supermin it, but it's unfortunate because we don't actually need privileges.

Reply:
Yeah. Hmm... maybe we could decouple the cachedir a bit more so it's available over 9p and so not part of the qcow2 while the workdir would still be on a POSIX fs in the VM. That will make composes slightly slower though because we'd no longer be able to use devino caching. But it'd allow e.g. `cosa fetch` and this code to run fully unprivileged. Not sure, probably not worth the trouble short-term.